File: backend/models/bet.go
package models

import "math"

// Match represents a bet made by one player
type Bet struct {
	Match              Match
	PredictedHomeGoals int
	PredictedAwayGoals int
}

// NewBet creates a new Bet instance
func NewBet(match Match, predictedHomeGoals int, predictedAwayGoals int) *Bet {
	return &Bet{
		Match:              match,
		PredictedHomeGoals: predictedHomeGoals,
		PredictedAwayGoals: predictedAwayGoals,
	}
}

// IsBetCorrect only checks if the predicted team won when the bet isn't a draw, or checks the draw the draw if it was predicted
func (b *Bet) IsBetCorrect() bool {
	if b.Match.GetHomeGoals() > b.Match.GetAwayGoals() {
		return b.PredictedHomeGoals > b.PredictedAwayGoals
	}
	if b.Match.GetHomeGoals() < b.Match.GetAwayGoals() {
		return b.PredictedHomeGoals < b.PredictedAwayGoals
	}
	return b.PredictedHomeGoals == b.PredictedAwayGoals
}

func (b *Bet) IsBetPerfect() bool {
	return b.PredictedHomeGoals == b.Match.GetHomeGoals() && b.PredictedAwayGoals == b.Match.GetAwayGoals()
}

func (b *Bet) AbsoluteDifferenceGoalDifferenceWithMatch() int {
	return int(math.Abs(float64(b.Match.AbsoluteGoalDifference() - b.AbsoluteGoalDifference())))
}

func (b *Bet) IsGoalDifferenceTheSameAsMatch() bool {
	return b.AbsoluteDifferenceGoalDifferenceWithMatch() == 0
}

func (b *Bet) AbsoluteGoalDifference() int {
	return int(math.Abs(float64(b.PredictedHomeGoals - b.PredictedAwayGoals)))
}

func (b *Bet) TotalPredictedGoals() int {
	return b.PredictedHomeGoals + b.PredictedAwayGoals
}

func (b *Bet) AbsoluteDifferenceTotalGoalsWithMatch() int {
	return int(math.Abs(float64(b.TotalPredictedGoals() - b.Match.TotalGoals())))
}

func (b *Bet) GetPredictedResult() string {
	if b.PredictedHomeGoals > b.PredictedAwayGoals {
		return b.Match.GetHomeTeam()
	}
	if b.PredictedHomeGoals < b.PredictedAwayGoals {
		return b.Match.GetAwayTeam()
	}
	return "Draw"
}

File: frontend/ligain/src/types/bet.ts
import { SeasonMatch, SimplifiedBet } from './match';

export interface Bet extends SimplifiedBet {
    match: SeasonMatch;
    isBetCorrect(): boolean;
    isBetPerfect(): boolean;
    absoluteDifferenceGoalDifferenceWithMatch(): number;
    isGoalDifferenceTheSameAsMatch(): boolean;
    absoluteGoalDifference(): number;
    totalPredictedGoals(): number;
    absoluteDifferenceTotalGoalsWithMatch(): number;
    getPredictedResult(): string;
}

export class BetImpl implements Bet {
    constructor(
        public match: SeasonMatch,
        public predictedHomeGoals: number,
        public predictedAwayGoals: number
    ) {}

    isBetCorrect(): boolean {
        if (this.match.getHomeGoals() > this.match.getAwayGoals()) {
            return this.predictedHomeGoals > this.predictedAwayGoals;
        }
        if (this.match.getHomeGoals() < this.match.getAwayGoals()) {
            return this.predictedHomeGoals < this.predictedAwayGoals;
        }
        return this.predictedHomeGoals === this.predictedAwayGoals;
    }

    isBetPerfect(): boolean {
        return this.predictedHomeGoals === this.match.getHomeGoals() && 
               this.predictedAwayGoals === this.match.getAwayGoals();
    }

    absoluteDifferenceGoalDifferenceWithMatch(): number {
        return Math.abs(this.match.absoluteGoalDifference() - this.absoluteGoalDifference());
    }

    isGoalDifferenceTheSameAsMatch(): boolean {
        return this.absoluteDifferenceGoalDifferenceWithMatch() === 0;
    }

    absoluteGoalDifference(): number {
        return Math.abs(this.predictedHomeGoals - this.predictedAwayGoals);
    }

    totalPredictedGoals(): number {
        return this.predictedHomeGoals + this.predictedAwayGoals;
    }

    absoluteDifferenceTotalGoalsWithMatch(): number {
        return Math.abs(this.totalPredictedGoals() - this.match.totalGoals());
    }

    getPredictedResult(): string {
        if (this.predictedHomeGoals > this.predictedAwayGoals) {
            return this.match.getHomeTeam();
        }
        if (this.predictedHomeGoals < this.predictedAwayGoals) {
            return this.match.getAwayTeam();
        }
        return 'Draw';
    }

    isModifiable(now: Date): boolean {
        return !this.match.isFinished() && !this.match.isInProgress() && now < this.match.getDate();
    }

    static fromJSON(json: any, match: SeasonMatch): BetImpl {
        return new BetImpl(
            match,
            json.predictedHomeGoals,
            json.predictedAwayGoals
        );
    }

    toJSON(): any {
        return {
            matchId: this.match.id(),
            predictedHomeGoals: this.predictedHomeGoals,
            predictedAwayGoals: this.predictedAwayGoals
        };
    }
} 